File: HW2/disambiguation.py
#!/usr/bin/env python
# encoding: utf-8

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# disambiguation.py: This script disambiguates a few word pairs using
#   either a Naive Bayes classifier or [EDIT] classifier.
#
# Author: April Walker
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
import numpy as np
import pandas as pd
import ipdb
import math
import time
import nltk
from nltk.corpus import stopwords as stopwords

stop_words = set(stopwords.words('english'))

file = open('amazon_reviews.txt', 'r', encoding='utf-8')
raw = file.read()

#
# save_to_file: simple function saves array of text to file with new line
#   between elements, snagged from interwebs and changed a tad

#
#   input--
#   filename: contains output filename/path (str)
#   text: array of text to write to file (array)
def save_to_file(filename, text):
    with open(filename, mode='wt', encoding='utf-8') as myfile:
        for line in text:
            myfile.write(line)
            myfile.write('\n')

#
# clean_corpus: minimally preprocesses corpus per nltk parser given by S. Gauch
#   input--
#   corpus: raw text to parse (str)
#
#   output--
#   words: an array of strings conatining the words in the corpus (list)
def clean_corpus(corpus):
    # just in case, this checks all chars are ascii
    corpus = ''.join(char for char in corpus if ord(char)<128)
    # tokenize into words
    tokens = nltk.word_tokenize(corpus)
    # make all chars lowercase, remove non-alpha chars
    words = [word.lower() for word in tokens if word.isalpha()]
    # get frequency of words for below operation
    freq = nltk.FreqDist(words)
    # remove words only occuring once
    words = [word for word in words if freq[word] > 1]
    return words
#
# wordpair_context: grabs conext and saves train/test from corpus for wordpairs
#   input--
#   wordpairs: contains all words for which context is wanted (array of tuples)
#   words: an array of strings conatining the words in the corpus (list)
#   n: gives the width for which context is wanted (int)
def wordpair_context(wordpairs, words, n):
    # get context: internal function to get context for specific word
    # inputs: my_word (str), words (array), n (int)
    # outputs: context (array)
    def get_context(my_word, words, n):
        # note word is skipped if appropropriate context not avaliable
        context = []
        for i in range(n, (len(words)-(n-1))):
            if words[i] == my_word:
                # get context for instance of word
                tmp = words[i-n:i+(n+1)]
                tmp.pop(n)
                # append context for this instance to array
                context.append(' '.join(tmp))
        return context

    for pair in wordpairs:
        for my_word in pair:
            # declare filenames
            train_file = '_'.join([my_word, "training.txt"])
            test_file = '_'.join([my_word, "testing.txt"])
            # get context
            context = get_context(my_word, words, n)
            # train/test split
            msk = np.random.rand(len(context)) < 0.8
            train = np.array(context)[msk]
            test = np.array(context)[~msk]
            # save train/test sets
            save_to_file(train_file, train)
            save_to_file(test_file, test)






# # # # # # # # # # # # # # # #
### Main Script Starts Here ###
# # # # # # # # # # # # # # # #

# # # Clean the corpus
print('Beginning preprocessing...')
start = time.time()
words = clean_corpus(raw)
end = time.time()
print((end - start), 'seconds elapsed')

# # # Step 1: Extract and save wordpairs

# indicate wordpairs of interest here
wordpairs = [("night", "seat"),
             ("kitchen", "cough"),
             ("car", "bike"),
             ("manufacturer", "bike"),
             ("big", "small"),
             ("huge", "heavy")]

# extract and save (only call this once...)
wordpair_context(wordpairs, words, 10)

for pair in wordpairs:
    # # #
    # Step 2: Prep out train/test datasets
    # # #
    train = pd.DataFrame()
    test = pd.DataFrame()
    truth = pd.DataFrame() # a copy of the test dataset with the true target
    for word in pair:
        # declare filenames to grab context
        train_file = '_'.join([word, "training.txt"])
        test_file = '_'.join([word, "testing.txt"])
        # get context
        new_train = pd.read_csv(train_file, delimiter=' ', header=None)
        new_test = pd.read_csv(test_file, delimiter=' ', header=None)
        # append true target value
        new_train["target"] = word
        new_test["target"] = word
        # append dataframes to train/truth sets
        train = train.append(new_train)
        truth = truth.append(new_test)
    # shuffle our dataframes once both wordpairs are present
    train = train.sample(frac=1)
    truth = truth.sample(frac=1)
    # drop our known target for our test set
    test = truth.drop(["target"], axis=1)

    # # #
    # Step 3: Training our classifiers. On top our standard naive bayes I'm
    #   going to try out some wild card ML classifiers. Because I expect these
    #   to take longer, I'll only test them out on a few instances of my
    #   word pairs.
    # # #
    wildcard_wordpairs = [("car", "bike"),
                          ("big", "small"),
                          ("huge", "heavy")]


    # While unecessary, I will save the classifer output to a txt for
    # curiosities sake... It might be neat to see *where* my classifiers
    # performed bad/good.


# # # Bonus Section:

# Here I'll repeat the process but with smaller/larger windows

# Here I'll experiment with a balanced training set



# fin #

File: HW2/disambiguation.R
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# disambiguation.py: Welcome to my R script. I assume you came my python
#   script so we'll skip the chit chat. Your standard CS gods have no power
#   here, this is no man's land.
#
# Author: April Walker
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
library(h2o)

# indicate wordpairs of interest here
word_pairs <- list(c("night", "seat"),
                   c("kitchen", "cough"),
                   c("car", "bike"),
                   c("manufacturer", "bike"),
                   c("big", "small"),
                   c("huge", "heavy"))

# # #
# Step 1: Dataset setup - this loop is very similar to the one ~ line 125 in R
#   however all train/test data is instead saved in a big ole list
# # #
train <- test <- list() # no truth set used for this code

for(pair in word_pairs) {
  # declare train/test sets for this wordpair
  pair_name <- paste(pair, collapse="")
  train[[pair_name]] <- test[[pair_name]] <- list()

  for(word in pair) {
    # declare filenames to grab context
    train_file <- paste(word, "training.txt", sep="_")
    test_file <- paste(word, "testing.txt", sep="_")
    # get context
    new_train <- read.table(train_file, header=F)
    new_test <- read.table(test_file, header=F)
    # append true target value
    new_train["target"] = word
    new_test["target"] = word
    # append dataframes to train/truth sets
    train[[pair_name]] <- rbind(train[[pair_name]], new_train)
    test[[pair_name]] <- rbind(test[[pair_name]], new_test)
  }
  # shuffle those bois
  train[[pair_name]] <- train[[pair_name]][sample(nrow(train[[pair_name]])),]
  test[[pair_name]] <- test[[pair_name]][sample(nrow(test[[pair_name]])),]
}

# # #
# Step 2: Machine learning funtime! I'll probably declare a few algorithms tbh
# # #

# Initialize ML pacakage h2o
h2o.init()

for (pair in word_pairs) {
  # # # Step 2.1 Initialize h2o dataframes / gridnames
  pair_name <- paste(pair, collapse="")

  this_train <- train[[pair_name]]
  this_test <- test[[pair_name]]

  train.hex <- as.h2o(this_train)
  test.hex <- as.h2o(this_test)

  # set all cols as factors for categorical analysis
  train.hex <- as.factor(train.hex)
  test.hex <- as.factor(test.hex)

  # # # Step 2.2 Commence GBM Classifier

  # save grid search
  grid_name <- paste0("gbm_", pair_name)

  # GBM hyper parameters
  params <- list(learn_rate = c(0.01, 0.05, 0.1),
                 max_depth = c(3, 10, 30))
  # Call grid search
  grid <- h2o.grid("gbm", y = "target",
                          grid_id = grid_name,
                          training_frame = train.hex,
                          validation_frame = test.hex,
                          learn_rate_annealing = 0.99,
                          ntrees = 100,
                          seed = 1,
                          hyper_params = params)

  grid_perf <- h2o.getGrid(grid_id = grid_name,
                           sort_by = "accuracy",
                           decreasing = T)

  # Choose top performing by accuracy
  m <- h2o.getModel(grid_perf@model_ids[[1]])

  # Write confusion matrix of test results to output file
  sink("gbm_results.txt", append = T)
  cat('\n')
  h2o.confusionMatrix(m, test.hex)
  sink(NULL)

  # # # End GBM Classifier

  # # # Step 2.3 Commence Neural Network

  # save grid search
  grid_name <- paste0("nn_", pair_name)

  # NN hyper parameters
  params <- list(hidden = c(1, 2, 5),
                 epochs = c(3, 5, 10))
  # Call grid search
  grid <- h2o.grid("deeplearning", y = "target",
                          grid_id = grid_name,
                          training_frame = train.hex,
                          validation_frame = test.hex,
                          seed = 1,
                          hyper_params = params)

  grid_perf <- h2o.getGrid(grid_id = grid_name,
                           sort_by = "accuracy",
                           decreasing = T)

  # Choose top performing by accuracy
  m <- h2o.getModel(grid_perf@model_ids[[1]])

  # Write confusion matrix of test results to output file
  sink("nn_results.txt", append = T)
  cat('\n')
  h2o.confusionMatrix(m, test.hex)
  sink(NULL)

  # # # End Neural Network Classifier
}


# Shutdown h2o
h2o.shutdown(prompt=T)
